packages/gatsby/src/db/sanitize-node.ts

```diff
@@ -1,12 +1,43 @@
-const _ = require(`lodash`)
+import _ from "lodash"
+import { Node } from "gatsby"
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type ObjectOrArray = Record<string, any> | Array<any> | Node
+
+/**
+ * Removes undefined keys from and Object or Array
+ */
+const omitUndefined = <T extends ObjectOrArray>(data: T): T => {
+  const isPlainObject = _.isPlainObject(data)
+  if (isPlainObject) {
+    return _.pickBy(data, p => p !== undefined) as T
+  }
+
+  return data.filter(p => p !== undefined)
+}
+
+const isTypeSupported = (data: unknown): boolean => {
+  if (data === null) {
+    return true
+  }
+
+  const type = typeof data
+  return (
+    type === `number` ||
+    type === `string` ||
+    type === `boolean` ||
+    data instanceof Date
+  )
+}
 
 /**
  * Make data serializable
- * @param {(Object|Array)} data to sanitize
- * @param {boolean} isNode = true
- * @param {Set<string>} path = new Set
  */
-const sanitizeNode = (data, isNode = true, path = new Set()) => {
+export function sanitizeNode<T extends Node>(
+  data: T,
+  isNode: boolean = true,
+  path: Set<T> = new Set()
+): T | undefined {
```

Comment:
Maybe the initial call to this recursive function should be separated (fine from a perf pov) because _that_ call _must_ return an `IGatsbyNode` whereas recursive calls can, but can also return a plain object. Not sure how to go about typing those tbh since it's, from TS's pov, untyped arbitrary content.